jf_agent/jira_download.py

```diff
@@ -271,13 +261,63 @@ def download_worklogs(jira_connection, issue_ids):
     }
 
 
-def _users_by_letter(jira_connection):
-    jira_users = set()
+def _jira_user_key(user_dict, gdpr_active):
+    if gdpr_active:
+        return user_dict['accountId']
+    else:
+        return user_dict['key']
+
+
+def _search_all_users(jira_connection, gdpr_active):
+    jira_users = {}
+    start_at = 0
+    while True:
+        # Search for '%' or '@' so we get results from cloud or
+        # on-prem, the latter being what on-prem appears to use as
+        # a wildcard
+        users = _search_users(jira_connection, gdpr_active, query='%', start_at=start_at, include_inactive=True) or \
+                _search_users(jira_connection, gdpr_active, query='@', start_at=start_at, include_inactive=True)
+        if not users:
+            return list(jira_users.values())
+
+        jira_users.update({
+            _jira_user_key(u, gdpr_active): u
+            for u in users
+        })
+
+        start_at += len(users)
+        
+
+def _users_by_letter(jira_connection, gdpr_active):
+    jira_users = {}
     for letter in string.ascii_lowercase:
-        jira_users.update(jira_connection.search_users(f'{letter}.', maxResults=1000, includeInactive=True, includeActive=False))
-        jira_users.update(jira_connection.search_users(f'{letter}.', maxResults=1000, includeInactive=False, includeActive=True))
-    return jira_users
+        jira_users.update({
+            _jira_user_key(u, gdpr_active): u
+            for u in _search_users(jira_connection, gdpr_active, query=f'{letter}.', include_inactive=True, include_active=False)
+        })
+        jira_users.update({
+            _jira_user_key(u, gdpr_active): u
+            for u in _search_users(jira_connection, gdpr_active, query=f'{letter}.', include_inactive=False, include_active=True)
+        })
+    return list(jira_users.values())
+
+
+def _search_users(jira_connection, gdpr_active, query, start_at=0, max_results=1000, include_active=True, include_inactive=False):
+    if gdpr_active:
+        # jira_connection.search_users creates a query that is no longer accepted on
+        # GDPR-compliant Jira instances.  Construct the right query by hand
+        params = {
+            'startAt': start_at,
+            'maxResults': max_results,
+            'query': query,
+            'includeActive': include_active,
+            'includeInactive': include_inactive
+        }
+        return jira_connection._get_json('user/search', params)
```

Comment:
This seems like a nice thing we could push upstream, see [`search_users` upstream](https://github.com/pycontribs/jira/blob/f49b92e8f0482cf987ff6ed02834632d28fed4c8/jira/client.py#L2726-L2748) which looks quite similar. It looks like there is [an upstream bug](https://github.com/pycontribs/jira/issues/697) for dealing with this.